controller/tenant.go

```diff
@@ -337,6 +298,74 @@ func listNames(namespaces []*tenant.Namespace) []string {
 	return names
 }
 
+func (c *TenantController) getOrInitTenant(ctx context.Context, user *auth.User) (*tenant.Tenant, error) {
+	var dbTenant *tenant.Tenant
+	var err error
+	tenantRepository := c.tenantService.NewTenantRepository(user.ID)
+	// check if tenant already exists
+	if tenantRepository.Exists() {
+		dbTenant, err = c.getExistingTenant(ctx, user.ID, user.OpenShiftUsername)
+		if err != nil {
+			log.Error(ctx, map[string]interface{}{
+				"err":      err,
+				"tenantID": user.ID,
+			}, "retrieval of tenant entity from DB failed")
+			return nil, errors.NewNotFoundError("tenant", user.ID.String())
+		}
+	} else {
+		nsBaseName, err := tenant.ConstructNsBaseName(c.tenantService, environment.RetrieveUserName(user.OpenShiftUsername))
+		if err != nil {
+			log.Error(ctx, map[string]interface{}{
+				"err":         err,
+				"os_username": user.OpenShiftUsername,
+			}, "unable to construct namespace base name")
+			return nil, errors.NewInternalError(ctx, err)
+		}
+		// if does not exist then create a new tenant
+		dbTenant = &tenant.Tenant{
+			ID:         user.ID,
+			Email:      *user.UserData.Email,
+			OSUsername: user.OpenShiftUsername,
+			NsBaseName: nsBaseName,
+		}
+		err = tenantRepository.CreateTenant(dbTenant)
+		if err != nil {
+			log.Error(ctx, map[string]interface{}{
+				"err": err,
+			}, "unable to store tenant configuration")
+			return nil, err
+		}
+	}
+	return dbTenant, nil
+}
+
+func (c *TenantController) createNamespaces(ctx context.Context, envTypes []environment.Type, user *auth.User, dbTenant *tenant.Tenant) error {
+	clusterNsMapping, err := c.clusterService.GetUserClusterForType(ctx, user)
```

Comment:
why is this method called `GetUserClusterForType`? what is the `type` here?

Reply:
https://github.com/fabric8-services/fabric8-tenant/blob/develop/cluster/service.go#L34
I could maybe change it to `GetUserClusterForEnvType`

Reply:
but you only pass the context and user as args here, so why `ForType` or `ForEnvType` ?

Reply:
it returns a function `ForType` that returns cluster for the given environment type. To point here is that it was a preparation for a multi-cluster environment (when one env type can be associated with a different cluster), that's why such a mapping. In this case, it just returns the same cluster (the user cluster) for all env types. This function is then used in a generic way.